### ISSUE TITLE
docker build hangs on COPY --from previous-stage

### ISSUE THREAD

**Issue by Antiarchitect** [moby/moby#40624]
I have docker out of docker setup for my Gitlab Builds in K8s 1.16.7 on baremetal.
I have a complicated multi-stage build that hangs on copy directory from the previous stage seems like for >20minutes. Have similar two COPY --from other-stage that do not hang. Size of the directory is about 8Mb - not much.
strace:
bash-5.0# strace -p 108
strace: Process 108 attached
futex(0x33fbac8, FUTEX_WAIT_PRIVATE, 0, NULL
stucks soooo long...

**Output of `docker version`:**

```
 docker version
 Client: Docker Engine - Community
  Version:           19.03.7
  API version:       1.40
  Go version:        go1.12.17
  Git commit:        7141c199a2
  Built:             Wed Mar  4 01:19:42 2020
  OS/Arch:           linux/amd64
  Experimental:      false
 Server: Docker Engine - Community
  Engine:
   Version:          19.03.7
   API version:      1.40 (minimum version 1.12)
   Go version:       go1.12.17
   Git commit:       7141c199a2
   Built:            Wed Mar  4 01:19:50 2020
   OS/Arch:          linux/amd64
   Experimental:     false
  containerd:
   Version:          v1.2.13
   GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
  runc:
   Version:          1.0.0-rc10
   GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
  docker-init:
   Version:          0.18.0
   GitCommit:        fec3683
```

**Output of `docker info`:**

```
docker info
 WARNING: API is accessible on http://0.0.0.0:2375 without encryption.
          Access to the remote API is equivalent to root access on the host. Refer
          to the 'Docker daemon attack surface' section in the documentation for
          more information: https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface
 Client:
  Debug Mode: false
 Server:
  Containers: 0
   Running: 0
   Paused: 0
   Stopped: 0
  Images: 32
  Server Version: 19.03.7
  Storage Driver: overlay2
   Backing Filesystem: <unknown>
   Supports d_type: true
   Native Overlay Diff: true
  Logging Driver: json-file
  Cgroup Driver: cgroupfs
  Plugins:
   Volume: local
   Network: bridge host ipvlan macvlan null overlay
   Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
  Swarm: inactive
  Runtimes: runc
  Default Runtime: runc
  Init Binary: docker-init
  containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
  runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
  init version: fec3683
  Security Options:
   seccomp
    Profile: default
  Kernel Version: 5.3.13-1.el7.elrepo.x86_64
  Operating System: Alpine Linux v3.11 (containerized)
  OSType: linux
  Architecture: x86_64
  CPUs: 8
  Total Memory: 15.64GiB
  Name: production-app-gitlab-runners-docker-0
  ID: ACPC:SJLM:E5P5:5XMH:4V3K:77NZ:NBSD:MQBC:SDC2:EANJ:4BMN:6P2L
  Docker Root Dir: /var/lib/docker
  Debug Mode: false
  Registry: https://index.docker.io/v1/
  Labels:
  Experimental: false
  Insecure Registries:
   127.0.0.0/8
  Live Restore Enabled: false
  Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Kubernetes 1.16.7
Centos 7 with 5.3.13-1.el7.elrepo.x86_64 kernel
Baremetal
Cilium 1.7.0 as network

**Comment by thaJeztah** [comment]
@tonistiigi PTAL

**Comment by Antiarchitect** [comment]
I will provide any debug info you need

**Comment by tonistiigi** [comment]
Can you provide a reproducer? Is this with or without buildkit? If without then try with `DOCKER_BUILDKIT=1` for data-point.

**Comment by Antiarchitect** [comment]
read(5, "\0\0\4\10\0\0\0\0\0\0\0\0\7\0\0\10\6\0\0\0\0\0\2\4\20\20\t\16\7\7", 4096) = 30
futex(0xc0003a44c8, FUTEX_WAKE_PRIVATE, 1) = 1
read(5, 0xc000216000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
After 10-15 minutes

**Comment by Antiarchitect** [comment]
The build is so complicated, but I will try without DOCKER_BUILDKIT tomorrow

**Comment by tonistiigi** [comment]
So this one was using buildkit? If you get it to hang you can get a stacktrace by sending `sigusr` signal to `dockerd`. Also post daemon logs with debug enabled with it. That `starce` snippet does not help. Not even sure what process it is taken from.

**Comment by Antiarchitect** [comment]
```
time="2020-03-04T22:53:22.893047070Z" level=warning msg="grpc: addrConn.createTransport failed to connect to { 0  <nil>}. Err :connection error: desc = \"transport: Error while dialing only one connection allowed\". Reconnecting..." module=grpc
time="2020-03-04T23:06:28.618813124Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.4.31:2375->10.32.9.206:46454: write: connection reset by peer"
time="2020-03-04T23:06:33.041004591Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.4.31:2375->10.32.9.206:46438: write: connection reset by peer"
time="2020-03-04T23:07:20.832177017Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.4.31:2375->10.32.9.69:37768: write: connection reset by peer"
time="2020-03-04T23:07:25.556277579Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.4.31:2375->10.32.9.69:37730: write: connection reset by peer"
time="2020-03-04T23:07:31.608088878Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.4.31:2375->10.32.3.115:37464: write: connection reset by peer"
time="2020-03-04T23:07:35.669295597Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.4.31:2375->10.32.3.115:37468: write: connection reset by peer"
```
Logs from the dockerd without debug

**Comment by Antiarchitect** [comment]
```
time="2020-03-05T00:22:20.027671859Z" level=info msg="parsed scheme: \"\"" module=grpc
time="2020-03-05T00:22:20.027777078Z" level=info msg="scheme \"\" not registered, fallback to default scheme" module=grpc
time="2020-03-05T00:22:20.027804539Z" level=info msg="ccResolverWrapper: sending update to cc: {[{ 0  <nil>}] <nil>}" module=grpc
time="2020-03-05T00:22:20.027818228Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
```
latest logs

**Comment by thaJeztah** [comment]
`"Error while dialing only one connection allowed"`; looks like I ran into that error at times; https://github.com/moby/buildkit/issues/748

**Comment by Antiarchitect** [comment]
I've just reconfigured my Gitlab runners and have positive results. It seems like when two (or more) similar builds using one docker daemon these livelocks occur.
Previous configuration scheme: 4 Gitlab runners each one exclusively uses one of the 4 docker-dind containers but concurrency is set to 2. So there is possible that two builds with the same Dockerfile can be run simultaneously on one docker-dind instance.
Current config: 8 Gitlab runners each one exclusively uses one of the 8 docker-dind containers and concurrency is set to 1. No more parallel builds. The problem is gone according to my observations.

**Comment by Antiarchitect** [comment]
```
time="2020-03-07T20:11:40.607663008Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/73e1fdb3c0d16cc4f216309e6336df17cd3ff89c8439bff0b766025c2cfdd364/shim.sock" debug=false pid=28669
time="2020-03-07T20:12:11.243161201Z" level=info msg="shim reaped" id=73e1fdb3c0d16cc4f216309e6336df17cd3ff89c8439bff0b766025c2cfdd364
time="2020-03-07T20:12:11.250361412Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:12:13.273913344Z" level=info msg="shim reaped" id=7cc5b1258e13569952deb3601b7120c113462258f8b1cba2282d7c096a512fcd
time="2020-03-07T20:12:13.278262123Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:12:13.278359468Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.3.53:2375->10.32.8.59:54826: write: connection reset by peer"
time="2020-03-07T20:12:18.066975844Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.3.53:2375->10.32.8.59:54818: write: connection reset by peer"
time="2020-03-07T20:12:18.184689957Z" level=info msg="shim reaped" id=bfab68b85c6bb99a6d77085ef5369c6e821f3753b8d1259532671b189de56d36
time="2020-03-07T20:12:18.194705442Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:28:16.886729079Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/7e833c1c0aae8c6e76c08338616c2fd403f0183ab7b7a94f92b1e71bf3b08824/shim.sock" debug=false pid=29647
time="2020-03-07T20:28:16.888276557Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/37eb8d95d2a17d6306ce8ad13f885e2be6e010d8d9f90848135f5c698963bf7b/shim.sock" debug=false pid=29648
time="2020-03-07T20:28:17.671522750Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/a14f5da9d1351dbdcd660a1bcf738ad113bb6f1e6ebcc6957553c28bc627c832/shim.sock" debug=false pid=29932
time="2020-03-07T20:28:45.632317976Z" level=info msg="shim reaped" id=a14f5da9d1351dbdcd660a1bcf738ad113bb6f1e6ebcc6957553c28bc627c832
time="2020-03-07T20:28:45.643889765Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:28:47.197027724Z" level=info msg="shim reaped" id=7e833c1c0aae8c6e76c08338616c2fd403f0183ab7b7a94f92b1e71bf3b08824
time="2020-03-07T20:28:47.208330171Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:28:47.208548647Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.3.53:2375->10.32.8.107:54314: write: connection reset by peer"
time="2020-03-07T20:28:52.062140816Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.3.53:2375->10.32.8.107:54308: write: connection reset by peer"
time="2020-03-07T20:28:52.157799457Z" level=info msg="shim reaped" id=37eb8d95d2a17d6306ce8ad13f885e2be6e010d8d9f90848135f5c698963bf7b
time="2020-03-07T20:28:52.167967951Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:29:07.686684637Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/68892bfe698488b03471b0de24073ece76123b346254a97d97bf915b0dfecd3c/shim.sock" debug=false pid=30888
time="2020-03-07T20:29:07.716994638Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/b90e99337da4312cece103db26cebb58fb23328a43ede43ac1753b29217b9fcd/shim.sock" debug=false pid=30903
time="2020-03-07T20:29:09.001581748Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/fd7800b9f675e8278065548cba432ba64989c300215a7bb58ee0f532d66c0b6e/shim.sock" debug=false pid=31202
time="2020-03-07T20:29:45.510042504Z" level=info msg="shim reaped" id=fd7800b9f675e8278065548cba432ba64989c300215a7bb58ee0f532d66c0b6e
time="2020-03-07T20:29:45.513254734Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:29:47.663175688Z" level=info msg="shim reaped" id=68892bfe698488b03471b0de24073ece76123b346254a97d97bf915b0dfecd3c
time="2020-03-07T20:29:47.672929577Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:29:47.673088638Z" level=error msg="attach failed with error: error attaching stderr stream: write tcp 10.32.3.53:2375->10.32.8.215:41102: write: connection reset by peer"
time="2020-03-07T20:29:52.166242332Z" level=error msg="attach failed with error: error attaching stdout stream: write tcp 10.32.3.53:2375->10.32.8.215:41094: write: connection reset by peer"
time="2020-03-07T20:29:52.266115970Z" level=info msg="shim reaped" id=b90e99337da4312cece103db26cebb58fb23328a43ede43ac1753b29217b9fcd
time="2020-03-07T20:29:52.276421191Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
time="2020-03-07T20:38:13.331568004Z" level=info msg="Attempting next endpoint for pull after error: manifest unknown: manifest unknown"
time="2020-03-07T20:39:08.920925219Z" level=info msg="parsed scheme: \"\"" module=grpc
time="2020-03-07T20:39:08.921002744Z" level=info msg="scheme \"\" not registered, fallback to default scheme" module=grpc
time="2020-03-07T20:39:08.921029430Z" level=info msg="ccResolverWrapper: sending update to cc: {[{ 0  <nil>}] <nil>}" module=grpc
time="2020-03-07T20:39:08.921042936Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
```

**Comment by Antiarchitect** [comment]
Observations were insufficient. One of my Gitlab builds stuck for more than one hour on the same step and was shut down by the timeout. That was the 1 - 1 configuration and the docker daemon was used exclusively. If you tell me more thoroughly what debug info should I provide I will do my best.

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/issues/40624#issuecomment-594966121

Perhaps you're able to get a stack trace using `sudo kill -SIGUSR1 $(pidof dockerd)`, this will write the stack-trace to a file (and logs the location of the file in the daemon logs)

**Comment by Antiarchitect** [comment]
Here is the stack-trace: https://gist.github.com/Antiarchitect/efa2bee9d15872b932cb807380e88380

**Comment by thaJeztah** [comment]
Thanks! @tonistiigi ptal

**Comment by Antiarchitect** [comment]
BTW just a little change - I've updated docker and docker-dind to the latest 19.03.8

**Comment by Antiarchitect** [comment]
@tonistiigi Sorry for bothering, have you advanced on this?

**Comment by tonistiigi** [comment]
Nothing is clearly stuck in that trace. It is in the middle of doing a cache load. You can start the daemon with `BUILDKIT_SCHEDULER_DEBUG=1` and report back the logs. Note that this will generate a lot of data so clearly mark in the log when you started the build and when it seemed stuck to you.

**Comment by Antiarchitect** [comment]
@tonistiigi Done. As the dump may contain confidential data I've sent it to your email. The dump ends where the task was stuck.

**Comment by Antiarchitect** [comment]
And this is the gap when all this "unstuck":

```
time="2020-03-16T23:07:43.633354573Z" level=debug msg="Layer already exists: sha256:95527ba98c29e7c9d47947c548c1ad5e42199f24f4401740fff9d630dc3170e4"
time="2020-03-16T23:32:05.743972923Z" level=debug msg=">> unpark [stage-7 8/9] COPY --from=erp-assets /app/apps/erp/public /app/apps/erp/public req=1 upt=1 out=1 state=cache-slow sha256:d35eb1def645e6a1c42b159ae8b0f3e7c8c3625872d18e23ed7abcdce3326923"
```

**Comment by Antiarchitect** [comment]
I've just deployed our docker-dind farm with `-D` and `BUILDKIT_SCHEDULER_DEBUG=1` and now running some builds. Will inform you later about the results

**Comment by thaJeztah** [comment]
@tonistiigi I see a potential fix was merged in https://github.com/moby/buildkit/pull/1413 - is that one to backport to the 0.6.x branch?

**Comment by thaJeztah** [comment]
opened https://github.com/moby/buildkit/pull/1422 for consideration

**Comment by Antiarchitect** [comment]
@thaJeztah Is this issue fixed in 19.03.9? Cannot follow non-trivial merges chain :)

**Comment by thaJeztah** [comment]
The fix I mentioned has been backported to 19.03.9 through https://github.com/moby/moby/pull/40784. It was mentioned as a possible fix for this issue, so I'd suggest trying if the issue is resolved in 19.03.9. (keep us posted 🤗)

**Comment by Antiarchitect** [comment]
@thaJeztah So far so good! I haven't seen any problems during the day. If I notice something I will reopen it.
